cmd/proxy/udp_proxy.go

```diff
@@ -110,7 +162,20 @@ func (proxy *UDPProxy) replyLoop(proxyConn *net.UDPConn, clientAddr *net.UDPAddr
 func (proxy *UDPProxy) Run() {
 	readBuf := make([]byte, UDPBufSize)
 	for {
-		read, from, err := proxy.listener.ReadFromUDP(readBuf)
+		// Use oob data/ControlMessages to get the dst IP of the
+		// received packet. This is used to ensure the src IP of the
+		// response matches when bound to 0.0.0.0 on  multi-homed
+		// machines.
+		var oobBuf []byte
+
+		if proxy.inet6Socket {
+			oobBuf = ipv6.NewControlMessage(ipv6.FlagDst | ipv6.FlagInterface)
+		} else {
+			oobBuf = ipv4.NewControlMessage(ipv4.FlagDst)
```

Comment:
Seems like we have the same issue here as we did on `Write` where we allocate a new buffer for every read even though the buffere size should be constant.